src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -166,6 +203,52 @@ snc_quota_manager::get_quota_balancer_node_period() const {
     return v;
 }
 
+void snc_quota_manager::update_node_quota_default() {
+    const inoutpair<std::optional<quota_t>> new_node_quota_default
+      = calc_node_quota_default();
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        // downstream updates:
+        // - shard effective quota (via _shard_quotas_update):
+        to_each(
+          [](
+            collect_dispense_amount_t& shard_quotas_update,
+            const std::optional<quota_t>& qold,
+            const std::optional<quota_t>& qnew) {
+              collect_dispense_amount_t diff;
+              if (qold && qnew) {
+                  diff.amount = *qnew - *qold;
+                  diff.amount_is = collect_dispense_amount_t::delta;
+              } else if (qold || qnew) {
+                  diff.amount = node_to_shard_quota(qnew);
+                  diff.amount_is = collect_dispense_amount_t::value;
+              }
+              shard_quotas_update.append(diff);
+          },
+          _shard_quotas_update,
+          _node_quota_default,
+          new_node_quota_default);
+    }
+    _node_quota_default = new_node_quota_default;
+    // - shard minimum quota:
+    update_shard_quota_minimum();
+    // apply the configuration change to the balancer
+    notify_quota_balancer_of_config_change();
+}
+
+void snc_quota_manager::update_shard_quota_minimum() {
+    _shard_quota_minimum = to_each(
+      [this](const std::optional<quota_t>& node_quota_default) {
+          const auto shard_quota_default = node_to_shard_quota(
+            node_quota_default);
+          return std::max<quota_t>(
+            _kafka_quota_balancer_min_shard_thoughput_ratio()
+              * shard_quota_default,
+            _kafka_quota_balancer_min_shard_thoughput_bps());
+      },
+      _node_quota_default);
+    // downstream updates: none
```

Comment:
`to_each` is still unintuitive to me. is this effectively the same thing as:

```c++
auto f = [this](p) {
          const auto shard_quota_default = node_to_shard_quota(p);
          return std::max<quota_t>(
            _kafka_quota_balancer_min_shard_thoughput_ratio()
              * shard_quota_default,
            _kafka_quota_balancer_min_shard_thoughput_bps());
};
_shard_quota_minimum.in = f(_node_quota_default.in);
_shard_quota_minimum.out = f(_node_quota_default.out);
```

Reply:
Exactly. The name `to_each` is a shorthand from "apply the function to each side of every argument, and combine the results".

Reply:
it kind of seems like this is a perfect case to use c++ ranges or boost ranges. almost all your ranges would have length 2, but a few cases it looks like you might have range lenght 4.

Reply:
refactored